test/tbbexec/test_tbb_thread_pool.cpp

```diff
@@ -0,0 +1,246 @@
+/*
+ * Copyright (c) 2022 Lucian Radu Teodorescu
+ * Copyright (c) 2022 NVIDIA Corporation
+ *
+ * Licensed under the Apache License Version 2.0 with LLVM Exceptions
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *   https://llvm.org/LICENSE.txt
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <catch2/catch.hpp>
+#include <stdexec/execution.hpp>
+#include <test_common/schedulers.hpp>
+#include <test_common/receivers.hpp>
+#include <exec/on.hpp>
+#include <tbbexec/tbb_thread_pool.hpp>
+
+#include <iostream>
+
+#include <map>
+
+namespace ex = stdexec;
+
+template <ex::scheduler Sched = inline_scheduler>
+inline auto _with_scheduler(Sched sched = {}) {
+  return exec::write(exec::with(ex::get_scheduler, std::move(sched)));
+}
+
+template <ex::scheduler Sched = inline_scheduler>
+inline auto _make_env_with_sched(Sched sched = {}) {
+  return exec::make_env(exec::with(ex::get_scheduler, std::move(sched)));
+}
+
+using _env_with_sched_t = decltype(_make_env_with_sched());
+
+TEST_CASE(
+    "exec::on works when changing threads with tbbexec::tbb_thread_pool", "[adaptors][exec::on]") {
+  tbbexec::tbb_thread_pool pool{2};
+  bool called{false};
+  // launch some work on the thread pool
+  ex::sender auto snd = exec::on(pool.get_scheduler(), ex::just()) //
+                        | ex::then([&] { called = true; }) | _with_scheduler();
+  stdexec::sync_wait(std::move(snd));
+  // the work should be executed
+  REQUIRE(called);
+}
+
+namespace {
+
+int compute(int x) { return x + 1; }
+
+std::size_t legible_thread_id() {
```

Comment:
I generally just print the std::hash of the thread ID.  That should save you from needing a mutex here.